### ISSUE TITLE
Быстрый поиск по оглавлению

### ISSUE THREAD

**Issue by alyuev** [tigran123/urantia-book-explorer#21]
Предлагаю добавить поле быстрого поиска по оглавлению в колонке "Оглавление". Фильтровать список по мере набора текста в этом поле и показывать в виде динамического меню. Т.е. его содержимое будет меняться по мере ввода символов в строке поиска.

Т.о. можно будет найти документ по номеру и быстро перейти к нему.

**Comment by tigran123** [comment]
Только не в колонке Оглавление! Эта колонка имеет структуру дерева и не надо её нарушать.

Но сама идея хорошая: я такое реализовал в графе "автор" в моей программе "Библиотека Урантии" (идея была создать планетарную библиотеку, даже межпланетную --- тестировал на 80-100 миллионов книг) --- там с помощью ajax происходил "autocompletion", т.е. быстро сканировались подходящие авторы и предоставлялись в виде списка.

Эту функцию можно добавить в нашу обычную строку поиска, но чтобы autocompletion включался только в определённых режимах (а не делал сотни ненужных поисков по всему тексту), скажем только по авторам документов или по названиям. Это можно контроллировать либо форматом самого ввода (например "author: бла-бла") или дополнительной кнопкой меняющей функцию поиска, т.е. сужающей её до множества легко-обозримого.

**Comment by alyuev** [comment]
А почему не в колонке "Оглавление"? Я имею в виду в шапке колонки оглавление.

**Comment by tigran123** [comment]
Т.е. вводишь в шапке оглавления скажем "Небадон" и дерево сразу заметно редеет, т.е. остаются только элементы, содержащие "Небадон"? Потому что структура будет нарушена --- например листья (т.е. секции/главы) дерева содержащие "Небадон" останутся, а промежуточные ветки по логике должны исчезнуть. А если их оставить, получится довольно "странное" дерево.

Кстати, у нас уже сейчас есть поиск "Заголовки". Ну вот, добавь туда autocompletion и будет тебе то, что ты хочешь, не ломая структуры данных хранящихся в столбце Оглавление.

**Comment by alyuev** [comment]
Нет, ты неправильно понял. Я указал, что по мере ввода текста сразу под полем ввода должно появляться динамическое меню, уже отфильтрованное. А само содержание не трогаем. Переход к нужной позиции в оглавлении будет делаться уже из этого динамического меню.

**Comment by tigran123** [comment]
Если содержание не трогаем, то это тоже самое, что я сказал выше, т.е. делай поиск в режиме "Заголовки", но с добавленной функции авто-заполнения. Для этого не надо дырявить шапку Оглавления. Есть уже окно ввода текста для поиска --- туда и вводи.

**Comment by alyuev** [comment]
В шапку оглавления добавляется всего лишь одно текстовое поле. Зато оно находится по месту. И им можно быстро воспользоваться. Часто, например, во время обсуждения кто-то говорит номер документа. И тогда ты должен: 1)выбрать тип поиска 2) удалить предыдущий текст поиска 3)ввести новый 4)перейти. А так мы сразу убираем первые два...

**Comment by tigran123** [comment]
Хорошо, подумаем, но пока что ты меня не убедил. Мне кажется что авто-заполнение обычной строки поиска в режиме "Заголовки" вполне справилось бы с той же функцией. Для быстрого перехода в конкретное место (по номеру/главе/параграфу) можно подправить поиск, чтобы "10:1.1" находил только 10:1.1, а не его и 110:1.1 в придачу.... Это было бы полезно --- мгновенный переход в любую точку текста без волокиты с Оглавлением или выбором вариантов из Результатов Поиска (и БЕЗ добавления особой панели специально для этого, как у Барри!)

**Comment by alyuev** [comment]
да, согласен, поиск по номеру/главе/параграфу тоже очень пригодится.

**Comment by tigran123** [comment]
Другое дело, что можно сделать то, что ты предлагаешь, НО тогда можно выкинуть наш селектор "Текст/Текст + заголовки/Только заголовки" вообще. Действительно, зачем он вообще нужен, если в столбце Оглавление будет специализированный поиск? Тогда мы не только не займём лишнего места, но даже освободим место, занимаемое ныне селектором. Я не против такого варианта.

**Comment by alyuev** [comment]
Можно и так. Было бы разумно. Тем более, если удастся реализовать и в новом месте сложный поиск по оглавлению.

**Comment by tigran123** [comment]
Чтобы можно было непосредственно прыгать в данное место текста нужно будет решить небольшую проблему: при печатании SRT спецификатора, например "47:10.6" он должен преобразовываться в форму: "U47_10_0". И то, это позволит перейти в начало главы, но не в данный параграф.

**Comment by tigran123** [comment]
Может быть стоит распознать в самом начале `search.php` форму запроса "X:Y.Z" (где Χ,Υ,Ζ целые числа) и быстренько найти это место (без всяких выкрутасов с запятыми и курсивами). Я подумаю над этим.

**Comment by tigran123** [comment]
Да, это нетрудно сделать и не нужно все файлы искать, т.е. номер документа сразу извлекается из SRT спецификатора и открывается нужный файл. Я сейчас это вариант имплементирую и будет у нас быстрый доступ по SRT ссылке.

**Comment by tigran123** [comment]
На всякий случай напомню, что после загрузки программы у нас все TOC уже загружены в память и при реализации поиска нам уже не нужно заново читать toc.html --- вообще к серверу доступ не нужен. Всё это должно быть сделано на jQuery: см. например как отображается название документа при помощи поиска по toc:

```
$('#' + col + 'title').html(toc.find('.U' + paper + '_0_1').html());
```

**Comment by tigran123** [comment]
И ещё: поиск не должен быть ограничен только названиями (документов и глав), но учитывать и остальные данные в нашей TOC структуре, а именно имя автора. Т.е. если я напечатаю "Divine Counselor" он должен показать список документов написанных этим существом.
